### PR TITLE
Add openGV support for rig localization

### PR DESCRIPTION
**Pull request by simogasp**
* [x] added OpenGV dependency
* [x] added localizeRig() implementation for all *Localizers
* [x] added unity tests
* [x] added travis rule to build and cache OpenGV
* [x] added `LocalizationResult` as output of `localizeRig()` connected to poparteu/openMVG#208

connected to poparteu/openMVG#93

### CODE REVIEW

File: src/openMVG/localization/CCTagLocalizer.cpp [hunk 1]
```diff
@@ -426,18 +428,65 @@ bool CCTagLocalizer::localizeRig(const std::vector<image::Image<unsigned char> >
...
+          
+  return localizeRig(vec_queryRegions,
+                     vec_imageSize,
+                     parameters,
+                     vec_queryIntrinsics,
+                     vec_subPoses,
+                     rigPose);
+}
+
+#ifndef HAVE_OPENGV
```
Comment (fabiencastan) [inline src/openMVG/localization/CCTagLocalizer.cpp]: `#ifdef HAVE_OPENGV`

File: src/openMVG/localization/CCTagLocalizer.cpp [hunk 2]
```diff
@@ -426,18 +428,65 @@ bool CCTagLocalizer::localizeRig(const std::vector<image::Image<unsigned char> >
...
+
+// subposes is n-1 as we consider the first camera as the main camera and the 
+// reference frame of the grid
+bool CCTagLocalizer::localizeRig(const std::vector<std::unique_ptr<features::Regions> > & vec_queryRegions,
```
Comment (fabiencastan) [inline src/openMVG/localization/CCTagLocalizer.cpp]: `localizeRig_opengv` inside opengv ifdef.
`localizeRig_naive` out of opengv ifdef.

And then:
```
bool CCTagLocalizer::localizeRig(const std::vector<std::unique_ptr<features::Regions> > & vec_queryRegions, XXX)
{
#ifndef HAVE_OPENGV
  localizeRig_opengv();
#else
  localizeRig_naive();
#endif
}
```

So the naive version is still available with opengv.

File: src/openMVG/localization/CCTagLocalizer.cpp [hunk 3]
```diff
@@ -426,18 +428,65 @@ bool CCTagLocalizer::localizeRig(const std::vector<image::Image<unsigned char> >
...
+                                 const std::vector<std::pair<std::size_t, std::size_t> > &imageSize,
+                                 const LocalizerParameters *parameters,
+                                 std::vector<cameras::Pinhole_Intrinsic_Radial_K3 > &vec_queryIntrinsics,
+                                 const std::vector<geometry::Pose3 > &vec_subPoses,
+                                 geometry::Pose3 &rigPose)
+{
+  const CCTagLocalizer::Parameters *param = static_cast<const CCTagLocalizer::Parameters *>(parameters);
+  if(!param)
+  {
+    // error!
```
Comment (fabiencastan) [inline src/openMVG/localization/CCTagLocalizer.cpp]: useless comment

File: src/openMVG/localization/CCTagLocalizer.cpp [hunk 4]
```diff
@@ -426,18 +428,65 @@ bool CCTagLocalizer::localizeRig(const std::vector<image::Image<unsigned char> >
...
+    throw std::invalid_argument("The parameters are not in the right format!!");
```
Comment (fabiencastan) [inline src/openMVG/localization/CCTagLocalizer.cpp]: "The CCTag localizer parameters are not in the right format."

File: src/openMVG/localization/CCTagLocalizer.cpp [hunk 5]
```diff
@@ -426,18 +428,65 @@ bool CCTagLocalizer::localizeRig(const std::vector<image::Image<unsigned char> >
...
+  }
+
+  const size_t numCams = vec_queryRegions.size();
+  
+  assert(numCams==vec_queryIntrinsics.size());
+  assert(numCams==vec_subPoses.size()-1);
```
Comment (fabiencastan) [inline src/openMVG/localization/CCTagLocalizer.cpp]: In another PR, it would be nice to remove the notion of main camera at the center of the world but have one subpose per camera in all case.

File: src/openMVG/localization/CCTagLocalizer.hpp [hunk 6]
```diff
@@ -78,34 +78,43 @@ class CCTagLocalizer : public ILocalizer
    * @return true if the rig has been successfully localized.
```
Comment (fabiencastan) [inline src/openMVG/localization/CCTagLocalizer.hpp]: vec_queryIntrinsics add => `(an initial guess is mandatory)`

File: src/openMVG/localization/CCTagLocalizer.cpp (hunk 5 above)
Comment (simogasp) [inline src/openMVG/localization/CCTagLocalizer.cpp]: H2020

File: src/openMVG/localization/CCTagLocalizer.cpp (hunk 2 above)
Comment (simogasp) [inline src/openMVG/localization/CCTagLocalizer.cpp]: done

File: src/openMVG/localization/CCTagLocalizer.cpp (hunk 4 above)
Comment (simogasp) [inline src/openMVG/localization/CCTagLocalizer.cpp]: done

File: src/openMVG/localization/CCTagLocalizer.cpp [hunk 7]
```diff
@@ -421,10 +422,11 @@ CCTagLocalizer::~CCTagLocalizer()
...
 // reference frame of the grid
 bool CCTagLocalizer::localizeRig(const std::vector<image::Image<unsigned char> > & vec_imageGrey,
-                              const LocalizerParameters *parameters,
-                              std::vector<cameras::Pinhole_Intrinsic_Radial_K3 > &vec_queryIntrinsics,
-                              const std::vector<geometry::Pose3 > &vec_subPoses,
-                              geometry::Pose3 &rigPose)
+                                 const LocalizerParameters *parameters,
+                                 std::vector<cameras::Pinhole_Intrinsic_Radial_K3 > &vec_queryIntrinsics,
+                                 const std::vector<geometry::Pose3 > &vec_subPoses,
+                                 geometry::Pose3 &rigPose,
```
Comment (fabiencastan) [inline src/openMVG/localization/CCTagLocalizer.cpp]: LocalizationResults already include poses, isn't it?

Comment (simogasp) [inline src/openMVG/localization/CCTagLocalizer.cpp]: yes, but in a far future the main camera won't be necessarily coincide with the pose of the rig

File: src/openMVG/localization/CCTagLocalizer.cpp [hunk 8]
```diff
@@ -532,14 +539,82 @@ bool CCTagLocalizer::localizeRig_opengv(const std::vector<std::unique_ptr<featur
...
+    for(const auto &ass : occurrences)
+    {
+      // recopy the associations IDs in the vector
+      indMatch3D2D.push_back(ass.first);
+    }
+    
+    vec_locResults.emplace_back(matchData, indMatch3D2D, pose, intrinsics, std::vector<voctree::DocMatch>(), refineOk);
+  }
+  
+  return resectionOk;
```
Comment (fabiencastan) [inline src/openMVG/localization/CCTagLocalizer.cpp]: `return refineOk;`

resectionOk will always be true here.

File: src/openMVG/localization/CCTagLocalizer.cpp (hunk 7 above)
Comment (fabiencastan) [inline src/openMVG/localization/CCTagLocalizer.cpp]: So we have to remove the `iLocalizer != 0` in refineRigPose... ;)

File: src/openMVG/rig/rig_BA_ceres.hpp [hunk 9]
```diff
@@ -14,10 +14,9 @@ namespace rig {
 class ResidualErrorMainCameraFunctor
 {
 public :
-  ResidualErrorMainCameraFunctor(
-          const cameras::Pinhole_Intrinsic_Radial_K3 & intrinsics,
-          const Vec2 & x,
-          const Vec3 & X)
+  ResidualErrorMainCameraFunctor(const cameras::Pinhole_Intrinsic_Radial_K3 & intrinsics,
+                                 const Vec2 & pt2d,
+                                 const Vec3 & pt3d)
```
Comment (fabiencastan) [inline src/openMVG/rig/rig_BA_ceres.hpp]: +1

File: src/openMVG/rig/Rig.cpp [hunk 10]
```diff
@@ -536,39 +535,39 @@ bool Rig::optimizeCalibration()
       _vPoses.push_back(pose);
     }
   }
-
-  displayRelativePoseReprojection(geometry::Pose3(openMVG::Mat3::Identity(), openMVG::Vec3::Zero()), 0);
-
-  // Update all poses over all witness cameras after optimization
+    
+    displayRelativePoseReprojection(geometry::Pose3(openMVG::Mat3::Identity(), openMVG::Vec3::Zero()), 0);
```
Comment (fabiencastan) [inline src/openMVG/rig/Rig.cpp]: indentation

File: src/openMVG/localization/CCTagLocalizer.cpp (hunk 7 above)
Comment (fabiencastan) [inline src/openMVG/localization/CCTagLocalizer.cpp]: This also means the vec_subPoses should contains all input poses:
`@param[in] vec_subPoses A vector containing the N-1 subposes of each camera wrt the main camera`

File: src/openMVG/localization/VoctreeLocalizer.cpp [hunk 11]
```diff
@@ -1038,13 +1040,410 @@ bool VoctreeLocalizer::robustMatching(matching::RegionsMatcherT<MatcherT> & matc
...
+    }
+    return resectionOk;
+  }
+  
+  { // just debugging stuff, this block can be removed
+    
+    if(vec_inliers.size() < numCams)
+    {
+      // in general the inlier should be spread among different cameras
+      POPART_CERR("****************** Oh, this is weird!!!!");
```
Comment (fabiencastan) [inline src/openMVG/localization/VoctreeLocalizer.cpp]: please add an explicit message:
```
"WARNING: RIG Voctree Localizer: Inliers in " << vec_inliers.size() << " cameras on a RIG of " << numCams << " cameras."
```

File: src/openMVG/localization/VoctreeLocalizer.cpp [hunk 12]
```diff
@@ -1038,13 +1040,410 @@ bool VoctreeLocalizer::robustMatching(matching::RegionsMatcherT<MatcherT> & matc
...
+    }
+    else
+    {
+      for(std::size_t cam = 0; cam < numCams; ++cam)
```
Comment (fabiencastan) [inline src/openMVG/localization/VoctreeLocalizer.cpp]: `for(std::size_t cam = 0; cam < vec_inliers.size(); ++cam)`
So you can always print it.

File: src/openMVG/localization/VoctreeLocalizer.cpp [hunk 13]
```diff
@@ -1038,13 +1040,410 @@ bool VoctreeLocalizer::robustMatching(matching::RegionsMatcherT<MatcherT> & matc
...
+    POPART_COUT("Rotation: " << rigPose.rotation());
+    POPART_COUT("Centre: " << rigPose.center());
+    
+    // compute the reprojection error for inliers (just debugging purposes)
+    for(std::size_t cam = 0; cam < numCams; ++cam)
+    {
+      const std::size_t numPts = vec_pts2D[cam].cols();
+      const cameras::Pinhole_Intrinsic_Radial_K3 &currCamera = vec_queryIntrinsics[cam];
+      Mat2X residuals;
+      if(cam!=0)
```
Comment (fabiencastan) [inline src/openMVG/localization/VoctreeLocalizer.cpp]: Could be removed if we take all local poses in input...

File: src/openMVG/localization/VoctreeLocalizer.cpp [hunk 14]
```diff
@@ -1038,13 +1040,410 @@ bool VoctreeLocalizer::robustMatching(matching::RegionsMatcherT<MatcherT> & matc
...
+#endif
+}
+
+
+#ifdef HAVE_OPENGV
+
+bool VoctreeLocalizer::localizeRig_opengv(const std::vector<std::unique_ptr<features::Regions> > & vec_queryRegions,
+                                          const std::vector<std::pair<std::size_t, std::size_t> > &vec_imageSize,
+                                          const LocalizerParameters *parameters,
+                                          std::vector<cameras::Pinhole_Intrinsic_Radial_K3 > &vec_queryIntrinsics,
```
Comment (fabiencastan) [inline src/openMVG/localization/VoctreeLocalizer.cpp]: Use boost::ptr_vector<IntrinsicBase> instead of std::vector<cameras::Pinhole_Intrinsic_Radial_K3 >.

File: src/openMVG/localization/CCTagLocalizer.cpp (hunk 7 above)
Comment (simogasp) [inline src/openMVG/localization/CCTagLocalizer.cpp]: yes, to be done on a separate PR,  see poparteu/openMVG#210

File: src/openMVG/localization/CCTagLocalizer.cpp (hunk 8 above)
Comment (simogasp) [inline src/openMVG/localization/CCTagLocalizer.cpp]: added `return true`

File: src/openMVG/rig/Rig.cpp (hunk 10 above)
Comment (simogasp) [inline src/openMVG/rig/Rig.cpp]: done

File: src/openMVG/localization/VoctreeLocalizer.cpp (hunk 13 above)
Comment (simogasp) [inline src/openMVG/localization/VoctreeLocalizer.cpp]: see poparteu/openMVG#210

File: src/openMVG/localization/VoctreeLocalizer.cpp (hunk 14 above)
Comment (simogasp) [inline src/openMVG/localization/VoctreeLocalizer.cpp]: see poparteu/openMVG#211

File: src/openMVG/localization/VoctreeLocalizer.cpp (hunk 12 above)
Comment (simogasp) [inline src/openMVG/localization/VoctreeLocalizer.cpp]: done

File: src/openMVG/localization/VoctreeLocalizer.cpp (hunk 11 above)
Comment (simogasp) [inline src/openMVG/localization/VoctreeLocalizer.cpp]: done

File: src/openMVG/localization/CCTagLocalizer.cpp (hunk 5 above)
Comment (simogasp) [inline src/openMVG/localization/CCTagLocalizer.cpp]: poparteu/openMVG#210

File: src/openMVG/localization/VoctreeLocalizer.cpp [hunk 15]
```diff
@@ -1204,11 +1204,9 @@ bool VoctreeLocalizer::localizeRig_opengv(const std::vector<std::unique_ptr<feat
               << vec_inliers.size() << " cameras on a RIG of " 
               << numCams << " cameras.");
     }
-    else
-    {
-      for(std::size_t cam = 0; cam < numCams; ++cam)
-        POPART_COUT("#inliers for cam " << cam << ": " << vec_inliers[cam].size());
-    }
+
+    for(std::size_t cam = 0; cam < numCams; ++cam)
```
Comment (fabiencastan) [inline src/openMVG/localization/VoctreeLocalizer.cpp]: numCams -> vec_inliers.size()

File: src/openMVG/rig/Rig.cpp [hunk 16]
```diff
@@ -536,7 +536,7 @@ bool Rig::optimizeCalibration()
     }
   }
     
-    displayRelativePoseReprojection(geometry::Pose3(openMVG::Mat3::Identity(), openMVG::Vec3::Zero()), 0);
+  displayRelativePoseReprojection(geometry::Pose3(openMVG::Mat3::Identity(), openMVG::Vec3::Zero()), 0);
     
     // Update all poses over all witness cameras after optimization
```
Comment (fabiencastan) [inline src/openMVG/rig/Rig.cpp]: same indentation problem around
